x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/ChunkedTrainedModelPersister.java

```diff
@@ -0,0 +1,235 @@
+import org.elasticsearch.xpack.ml.extractor.ExtractedFields;
+import org.elasticsearch.xpack.ml.extractor.MultiField;
+import org.elasticsearch.xpack.ml.inference.modelsize.ModelSizeInfo;
+import org.elasticsearch.xpack.ml.inference.persistence.TrainedModelDefinitionDoc;
+import org.elasticsearch.xpack.ml.inference.persistence.TrainedModelProvider;
+import org.elasticsearch.xpack.ml.notifications.DataFrameAnalyticsAuditor;
+
+import java.time.Instant;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+import static java.util.stream.Collectors.toList;
+
+public class ChunkedTrainedModelPersister {
+
+    private static final Logger LOGGER = LogManager.getLogger(ChunkedTrainedModelPersister.class);
+    private static final int STORE_TIMEOUT_SEC = 30;
+    private final TrainedModelProvider provider;
+    private final AtomicReference<String> currentModelId;
+    private final DataFrameAnalyticsConfig analytics;
+    private final DataFrameAnalyticsAuditor auditor;
+    private final Consumer<Exception> failureHandler;
+    private final ExtractedFields extractedFields;
+    private final AtomicBoolean readyToStoreNewModel = new AtomicBoolean(true);
+
+    public ChunkedTrainedModelPersister(TrainedModelProvider provider,
+                                        DataFrameAnalyticsConfig analytics,
+                                        DataFrameAnalyticsAuditor auditor,
+                                        Consumer<Exception> failureHandler,
+                                        ExtractedFields extractedFields) {
+        this.provider = provider;
+        this.currentModelId = new AtomicReference<>("");
+        this.analytics = analytics;
+        this.auditor = auditor;
+        this.failureHandler = failureHandler;
+        this.extractedFields = extractedFields;
+    }
+
+    public void createAndIndexInferenceModelDoc(TrainedModelDefinitionChunk trainedModelDefinitionChunk) {
+        if (Strings.isNullOrEmpty(this.currentModelId.get())) {
+            failureHandler.accept(ExceptionsHelper.serverError(
+                "chunked inference model definition is attempting to be stored before trained model configuration"
+            ));
+            return;
+        }
+        TrainedModelDefinitionDoc trainedModelDefinitionDoc = trainedModelDefinitionChunk.createTrainedModelDoc(this.currentModelId.get());
+
+        CountDownLatch latch = storeTrainedModelDoc(trainedModelDefinitionDoc);
+        try {
+            if (latch.await(STORE_TIMEOUT_SEC, TimeUnit.SECONDS) == false) {
+                LOGGER.error("[{}] Timed out (30s) waiting for chunked inference definition to be stored", analytics.getId());
```

Comment:
nit: now that `STORE_TIMEOUT_SEC` is a constant, the log message can use it as argument (in other places, too)